tests/test_config.py

```diff
@@ -227,3 +236,207 @@ def test_worker_config_precedence(tmp_path, test_input, expected_exclusions):
+            A list: four, five, six
+            """
+            ),
+        ),
+        # User defaults + data provided.
+        (
+            {
+                "a_string": "foo",
+                "a_number": 42,
+                "a_boolean": False,
+                "a_list": ["four", "five", "six"],
+            },
+            {
+                "a_string": "yosemite",
+            },
+            dedent(
+                """\
+            A string: yosemite
+            A number: 42
+            A boolean: False
+            A list: four, five, six
+            """
+            ),
+        ),
+    ],
+)
+def test_user_defaults(tmp_path_factory, user_defaults, data, expected):
+    src, dst = map(tmp_path_factory.mktemp, ("src", "dst"))
+    build_file_tree(
+        {
+            src
+            / "copier.yaml": """\
+                a_string:
+                    default: lorem ipsum
+                    type: str
+                a_number:
+                    default: 12345
+                    type: int
+                a_boolean:
+                    default: true
+                    type: bool
+                a_list:
+                    default:
+                        - one
+                        - two
+                        - three
+                    type: json
+            """,
+            src
+            / "user_data.txt.jinja": """\
+                A string: {{ a_string }}
+                A number: {{ a_number }}
+                A boolean: {{ a_boolean }}
+                A list: {{ ", ".join(a_list) }}
+            """,
+            src
+            / "{{ _copier_conf.answers_file }}.jinja": """\
+                # Changes here will be overwritten by Copier
+                {{ _copier_answers|to_nice_yaml }}
+            """,
+        }
+    )
+    copier.copy(
+        str(src),
+        dst,
+        defaults=True,
+        overwrite=True,
+        user_defaults=user_defaults,
+        data=data,
+    )
+    gen_file = dst / "user_data.txt"
+    result = gen_file.read_text()
+    assert result == dedent(expected)
+
+
+@pytest.mark.parametrize(
+    "user_defaults_initial, user_defaults_updated, data_initial, data_updated, expected",
+    [
+        # User defaults + secondary defaults.
+        (
+            {
+                "a_string": "foo",
+            },
+            {
+                "a_string": "foobar",
+            },
+            {},
+            {},
+            dedent(
+                """\
+            A string: foo
+            """
+            ),
+        ),
+        # User defaults + data provided.
+        (
+            {
+                "a_string": "foo",
+            },
+            {
+                "a_string": "foobar",
+            },
+            {
+                "a_string": "yosemite",
+            },
+            {
+                "a_string": "red rocks",
+            },
+            dedent(
+                """\
+            A string: yosemite
```

Comment:
If it passes it's weird. It should be this AFAICS:

```suggestion
            A string: red rocks
```

Reply:
Ack. Fixed the test in a subsequent test case:


```python
        (
            {
                "a_string": "foo",
            },
            {
                "a_string": "foobar",
            },
            {
                "a_string": "yosemite",
            },
            {
                "a_string": "red rocks",
            },
            dedent(
                """\
            A string: yosemite
            """
            ),
            dedent(
                """\
            A string: red rocks
            """
            ),
        ),
```

On the first pass we provide: 

```
user_default: foo
data: yosemite
```

result: yosemite

Second pass:

```
user_default: foobar
data: yosemite
```

result: red rocks (overriding the copier answers from the initial copy).